lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,516 @@
+    if (position != std::lower_bound(begin, end, subtag)) {
+      return false;
+    }
+    languageId.variantSubtagList.insert(position, subtag);
+  }
+  return true;
+}
+
+bool LanguageTagParser::parseExtensions() {
+  while (true) {
+    if (subtagEnd_ != subtagStart_) {
+      return true;
+    }
+    
+    char16_t singleton = getCurrentSubtag()[0];
+    if (!isASCIILetterOrDigit(singleton)) {
+      return true;
+    }
+    
+    if (!nextSubtag()) {
+      return false;
+    }
+    
+    // duplicate subtags handled in parse functions
+    switch (singleton) {
+      case 'u': {
+        // unicode extension
+        if (!parseUnicodeExtension()) {
+          return false;
+        }
+        break;
+      }
+      case 't': {
+        // transformed extension
+        if (!parseTransformedExtension()) {
+          return false;
+        }
+        break;
+      }
+      case 'x': {
+        // private use extension
+        if (!parsePUExtension()) {
+          return false;
+        }
+        return true;
+      }
+      default: {
+        // other extension
+        if (!parseOtherExtension(singleton)) {
+          return false;
+        }
+        break;
+      }
+    }
+  }
+
+  return false;
+}
+
+// unicode_locale_extensions = sep [uU]
+// ((sep keyword)+
+// |(sep attribute)+ (sep keyword)*) ;
+//
+// keyword = = key (sep type)? ;
+//
+// key = = alphanum alpha ;
+//
+// type = = alphanum{3,8}
+//  (sep alphanum{3,8})* ;
+//
+// attribute = alphanum{3,8} ;
+bool LanguageTagParser::parseUnicodeExtension() {
+  if (!parsedLocaleIdentifier.unicodeExtensionAttributes.empty() ||
+      !parsedLocaleIdentifier.unicodeExtensionKeywords.empty()) {
+    return false;
+  }
+  
+  while (isUnicodeExtensionAttribute(getCurrentSubtagRef())) {
+    // Insert in sorted order
+    if (parsedLocaleIdentifier.unicodeExtensionAttributes.empty()) {
+      parsedLocaleIdentifier.unicodeExtensionAttributes.push_back(getCurrentSubtag());
+    } else {
+      auto subtag = getCurrentSubtag();
+      auto begin = parsedLocaleIdentifier.unicodeExtensionAttributes.begin();
+      auto end = parsedLocaleIdentifier.unicodeExtensionAttributes.end();
+      auto position = std::upper_bound(begin, end, subtag);
+      parsedLocaleIdentifier.unicodeExtensionAttributes.insert(position, subtag);
+    }
+
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+  
+  while (isUnicodeExtensionKey(getCurrentSubtagRef())) {
+    std::u16string key = getCurrentSubtag();
+
+    parsedLocaleIdentifier.unicodeExtensionKeywords.insert({key, {}});
+    std::vector<std::u16string> *extensionKeyTypes = &parsedLocaleIdentifier.unicodeExtensionKeywords.find(key)->second;
+    
+    if (!nextSubtag()) {
+      return true;
+    } else {
+      while (isUnicodeExtensionKeyTypeItem(getCurrentSubtagRef())) {
+        extensionKeyTypes->push_back(getCurrentSubtag());
+        if (!nextSubtag()) {
+          return true;
+        }
+      }
+    }
+  }
+  
+  return true;
```

Comment:
From my reading of the spec, we must have at least one keyword or attribute, however, it is possible to get to this point without any.